haystack/document_stores/pinecone.py

```diff
@@ -435,22 +560,128 @@ def get_all_documents_generator(
+            else:
+                namespace = self.document_namespace
+
+        ids = self._get_all_document_ids(namespace=namespace, filters=filters, batch_size=batch_size)
+        for i in range(0, len(ids), batch_size):
+            i_end = min(len(ids), i + batch_size)
+            documents = self.get_documents_by_id(
+                ids=ids[i:i_end], namespace=namespace, batch_size=batch_size, return_embedding=return_embedding
+            )
+            for doc in documents:
+                yield doc
+
+    def _get_all_document_ids(
+        self,
+        index: Optional[str] = None,
+        namespace: Optional[str] = None,
+        filters: Optional[Dict[str, Union[Dict, List, str, int, float, bool]]] = None,
+        batch_size: int = 32,
+    ) -> List[str]:
+        index = str(index or self.index)
+        index = _sanitize_index_name(index)
+        if index not in self.pinecone_indexes:
+            raise DocumentStoreError(
+                f"Index named '{index}' does not exist. Try reinitializing PineconeDocumentStore() and running "
+                f"'update_embeddings()' to create and populate an index."
+            )
+
+        if namespace is None:
+            if self.get_embedding_count() > 0:
+                namespace = self.embedding_namespace
+            else:
+                namespace = self.document_namespace
+
+        document_count = self.get_document_count()
+        if len(self.all_ids) == document_count and filters is None:
+            # We have all of the IDs and don't need to extract from Pinecone
+            return list(self.all_ids)
+        else:
+            target_namespace = f"{namespace}-copy"
+            all_ids: Set[str] = set()
+            with tqdm(
+                total=document_count, disable=not self.progress_bar, position=0, unit=" ids", desc="Retrieving IDs"
+            ) as progress_bar:
+                while True:
+                    # Retrieve IDs from Pinecone
+                    vector_id_matrix = self._get_ids(
+                        index=index, namespace=namespace, batch_size=batch_size, filters=filters
+                    )
+                    # Once we reach final item, we break
+                    if len(vector_id_matrix) == 0:
+                        break
+                    # Save IDs
+                    all_ids = all_ids.union(set(vector_id_matrix))
+                    # Move these IDs to new namespace
+                    self._move_documents_by_id_namespace(
+                        ids=vector_id_matrix,
+                        source_namespace=namespace,
+                        target_namespace=target_namespace,
+                        batch_size=batch_size,
+                    )
+                    progress_bar.set_description_str("Retrieved IDs")
+                    progress_bar.update(batch_size)
+            # Now move all documents back to source namespace
+            self._namespace_cleanup(index)
```

Comment:
:question: Why a search method? Don't we know exactly the namespace name here?

Reply:
I think it would be possible another copy namespace to be created (for example "no-vectors-copy"), and if for example the user loses connection or quits the process, it can leave a "-copy" namespace in existence. If I user then uses this method for the "vectors" namespace the cleanup function will also find the older "no-vectors-copy" and move that back to the "no-vectors" namespace.

So not necessary, but I think it can be helpful, however I'm happy to change it if needed

Reply:
Oh right, you've got a point. If something occurs and the call breaks those namespaces can be left around. Looks like a decent workaround for now, let's keep it :+1: Let's just add a comment to clarify, so I won't accidentally delete this as soon as I forget about this conversation :smile: